### PR TITLE
Revamp wait package

### PR DESCRIPTION
**Pull request by hcsa73**
- Add generic to signature of wait function
- Rename types and fields, add comments
- Uniformize `SetXXX` signatures
- Rewrite tests

### CODE REVIEW

File: examples/waiter/waiter.go [hunk 1]
```diff
@@ -38,20 +38,13 @@ func main() {
 	}
 
 	zoneId := *createZoneResp.Zone.Id
+
+	// The following will wait until the DNS zone is finshed being created
```
Comment (joaopalet) [inline examples/waiter/waiter.go:42]:
```suggestion
	// The following will wait until the DNS zone creation has finished
```

File: core/wait/wait.go [hunk 2]
```diff
@@ -12,72 +12,77 @@ import (
 
 var RetryHttpErrorStatusCodes = []int{http.StatusBadGateway, http.StatusGatewayTimeout}
 
-type WaitFn func() (res interface{}, done bool, err error)
-
-type Handler struct {
-	fn                WaitFn
+// AsyncActionCheck reports whether a specific async action has finished.
+//   - waitFinished == true if the async action is finished, false otherwise.
+//   - response contains data regarding the current state of the resource targeted by the async action (if applicable). resource != nil if waitFinished == true.
```
Comment (joaopalet) [inline core/wait/wait.go:17]: > resource != nil if waitFinished == true.

Did you mean `resource != nil if (...)`? Maybe we have cases where `resource` is `nil` even though the wait has finished, so we could rephrase it to `response == nil if waitFinished == false`.

File: core/wait/wait.go [hunk 3]
```diff
@@ -12,72 +12,77 @@ import (
...
 }
 
-// SetThrottle sets the duration between func triggering
-func (w *Handler) SetThrottle(d time.Duration) error {
-	if d == 0 {
-		return fmt.Errorf("throttle can't be 0")
-	}
-	w.throttle = d
-	return nil
+// SetThrottle sets the duration between func triggering.
```
Comment (joaopalet) [inline core/wait/wait.go:41]: Maybe we could make this more explicit? Something like

```suggestion
// SetThrottle sets the time interval between each checkFn triggering.
```

File: core/wait/wait.go [hunk 4]
```diff
@@ -12,72 +12,77 @@ import (
...
+type AsyncActionCheck[T any] func() (waitFinished bool, response *T, err error)
+
+// AsyncActionHandler handles waiting for a specific async action to be finished.
+type AsyncActionHandler[T any] struct {
+	checkFn           AsyncActionCheck[T]
 	sleepBeforeWait   time.Duration
 	throttle          time.Duration
 	timeout           time.Duration
-	tempErrRetryLimit int
+	retryLimitTempErr int
```
Comment (joaopalet) [inline core/wait/wait.go:27]: I think this read a bit better as it was before (we should also update in the `Set...` function)

File: core/wait/wait.go [hunk 5]
```diff
@@ -12,72 +12,77 @@ import (
...
 
-type WaitFn func() (res interface{}, done bool, err error)
-
-type Handler struct {
-	fn                WaitFn
+// AsyncActionCheck reports whether a specific async action has finished.
+//   - waitFinished == true if the async action is finished, false otherwise.
+//   - response contains data regarding the current state of the resource targeted by the async action (if applicable). resource != nil if waitFinished == true.
+//   - err != nil if there was an error checking if the async action finished, or if it finished unsuccessfully.
+type AsyncActionCheck[T any] func() (waitFinished bool, response *T, err error)
```
Comment (joaopalet) [inline core/wait/wait.go:19]: What was the reasoning behind switching the output order?

File: core/wait/wait_test.go [hunk 6]
```diff
@@ -8,66 +8,135 @@ import (
...
 
 	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
 	oapiError "github.com/stackitcloud/stackit-sdk-go/core/oapierror"
 )
 
+// Options used for comparing AsyncActionHandler
+var cmpOpts = []cmp.Option{
+	cmp.AllowUnexported(AsyncActionHandler[interface{}]{}),
+	cmpopts.IgnoreFields(AsyncActionHandler[interface{}]{}, "checkFn"),
```
Comment (joaopalet) [inline core/wait/wait_test.go:18]: Why are we ingoring the `checkFn` field?

File: core/wait/wait_test.go [hunk 7]
```diff
@@ -8,66 +8,135 @@ import (
...
 			}
-			if got != nil && tt.want != nil {
-				if got.Error() != tt.want.Error() {
-					t.Errorf("Wait.SetThrottle() = %v, want %v", got, tt.want)
-				}
+			if got := w.SetThrottle(tt.args.d); !cmp.Equal(got, tt.want, cmp.AllowUnexported(AsyncActionHandler[interface{}]{})) {
+				t.Errorf("Wait.SetThrottle() = %v, want %v", got, tt.want)
 			}
 		})
 	}
```
Comment (joaopalet) [inline core/wait/wait_test.go:98]: Can be removed (same on below tests)

File: core/wait/wait_test.go [hunk 8]
```diff
@@ -169,25 +267,239 @@ func TestSetTimeout(t *testing.T) {
...
+
+			if tt.wantErr && (err == nil) {
+				t.Errorf("expected error but got none")
+			}
+			if !tt.wantErr && (err != nil) {
+				t.Errorf("expected no error but got \"%v\"", err)
+			}
+			if (err == nil) && (resp == nil) {
+				t.Errorf("got nil err but nil resp")
+			}
```
Comment (joaopalet) [inline core/wait/wait_test.go:493]: Similar to a comment I made above, can't we have `checkFn`s returning nil? If the API doesn't have a response body

File: core/wait/wait.go (hunk 2 above)
Comment (hcsa73) [reply core/wait/wait.go:17]: I did mean `resource != nil if waitFinished == true`.

It's technically possible for the function to not want to return anything when `waitFinished == true`, and the handler's `WaitWithContext` doesn't care if `resp` is `nil` or not. I'd say it's very unlikely that such an use case happens, though: for the function to decide that the wait is over, it must have done by using some API response, and so it can return that API response that proves the wait has finished (no matter if successful or unsuccessful).

I've checked, and not all usages of `AsyncActionHandler` this rule. Eg, `CreateLoadBalancerWaitHandler`, upon finding that the load balancer's creation failed, returns `true, nil, fmt.Errorf(...)`. However, it can return `true, s, fmt.Errorf(...)`, where `s` is the API response that holds the failed state of the API.

I'd suggest enforcing this rule, by 
- Checking if `resp != nil` in `WaitWithContext`
- Removing the `(if applicable)` portion of the comment
- Adding in the documentation of `WaitWithContext` that `resp != nil` if `err != nil`
- Fix usages of `AsyncActionHandler`

Comment (hcsa73) [reply core/wait/wait.go:17]: We can also enforce a weaker rule, that `resource != nil if waitFinished == true && err != nil`.

File: core/wait/wait.go (hunk 3 above)
Comment (hcsa73) [reply core/wait/wait.go:41]: Reworded it a bit

File: core/wait/wait.go (hunk 5 above)
Comment (hcsa73) [reply core/wait/wait.go:19]: Just that `waitFinished` is "more important", in the sense that this function gets called repeatedly until is equals `true`. Also `WaitWithContext`'s signature matches this one without the `waitFinished`, and it looked weird for it to be between the other two outputs

File: core/wait/wait_test.go (hunk 6 above)
Comment (hcsa73) [reply core/wait/wait_test.go:18]: Apparently `cmp` can't compare functions

File: core/wait/wait_test.go (hunk 7 above)
Comment (hcsa73) [reply core/wait/wait_test.go:98]: Ooops, leftover code

File: core/wait/wait_test.go (hunk 8 above)
Comment (hcsa73) [reply core/wait/wait_test.go:377]: Check comment above

File: core/wait/wait_test.go (hunk 6 above)
Comment (hcsa73) [reply core/wait/wait_test.go]: Added a comment for this

File: core/wait/wait.go (hunk 2 above)
Comment (hcsa73) [reply core/wait/wait.go:17]: After discussion, will remove the rule and keep it generic

**Comment by hcsa73** [comment]
Changed the rule for handlers for which the response part doesn't apply, uniformed implementation of Handlers and adapted tests

**Review by joaopalet — APPROVED** [review]
